File: approval_l1/scripts/generator.cpp
#include <algorithm>
#include <boost/program_options.hpp>
#include <chrono>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <nlohmann/json.hpp>
#include <vector>

#include "approvalwise_vector.hpp"
#include "brute.hpp"
#include "greedy_dp.hpp"

using json = nlohmann::json;
namespace po = boost::program_options;

signed main(int argc, char** argv) {
    po::options_description desc("Allowed options");
    desc.add_options()("help", "produce help message")(
        "voters,N", po::value<int>()->required(), "number of voters")(
        "candidates,M", po::value<int>()->required(), "number of candidates")(
        "additional,R", po::value<int>()->required(), "number of new votings")(
        "algorithm,A", po::value<std::string>()->default_value("brute"),
        "chosen algorithm <brute|greedy_dp>")(
        "json", po::value<bool>()->default_value(false), "json output?")(
        "verbose", po::value<int>()->default_value(1), "verbose level")(
        "load-hists", po::value<std::string>(), "path to json file with hists");

    po::variables_map vm;
    try {
        po::store(po::parse_command_line(argc, argv, desc), vm);
        po::notify(vm);
    } catch (const po::error& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    if (vm.count("help")) {
        std::cout << desc << "\n";
        return 1;
    }

    const int N = vm["voters"].as<int>();
    const int M = vm["candidates"].as<int>();
    const int R = vm["additional"].as<int>();
    const int verbose = vm["verbose"].as<int>();
    const bool json_output = vm["json"].as<bool>();
    const std::string algorithm = vm["algorithm"].as<std::string>();

    if (algorithm != "brute" && algorithm != "greedy_dp") {
        std::cerr << "Error: unknown algorithm" << std::endl;
        return 1;
    }

    vector<approvalwise_vector_t> votings_hist;
    if (vm.count("load-hists")) {
        std::ifstream ifs(vm["load-hists"].as<std::string>());
        json jsonData = json::parse(ifs);
        votings_hist = jsonData.get<vector<approvalwise_vector_t>>();
    } else {
        votings_hist = {vi(M, N), vi(M, 0)};
    }
    const size_t start_N = votings_hist.size();

    if (verbose >= 1) cout << "r,dist,dist_prop,time" << endl;
    if (json_output) {
        cout << "[\n";
        for (const auto& voting : votings_hist) {
            cout << "\t";
            print_vec(voting);
            cout << ",";
            cout << endl;
        }
    }

    for (int i = 0; i < R; i++) {
        auto start_time = chrono::high_resolution_clock::now();
        std::pair<approvalwise_vector_t, int> res;
        if (algorithm == "brute") {
            res = brute::farthest_approvalwise_vector(votings_hist, N);
        } else if (algorithm == "greedy_dp") {
            res = greedy_dp::farthest_approvalwise_vector(votings_hist, N);
        }
        auto [next_voting_hist, score] =
            brute::farthest_approvalwise_vector(votings_hist, N);

        auto end_time = chrono::high_resolution_clock::now();
        auto elapsed_time =
            chrono::duration_cast<chrono::milliseconds>(end_time - start_time)
                .count();

        votings_hist.push_back(next_voting_hist);
        if (verbose >= 1)
            cout << i + start_N + 1 << ',' << score << ','
                 << double(score) / (N * M) << ',' << setprecision(4)
                 << double(elapsed_time / 1000.) << endl;
        if (json_output) {
            cout << "\t";
            print_vec(next_voting_hist);
            if (i != R - 1) cout << ",";
            cout << endl;
        }
    }

    if (json_output) {
        cout << "]";
    }
}

File: approval_l1/scripts/generator.py
import gurobi
import basin_hopping as bh
import time
import numpy as np

import argparse

VotingHist = np.ndarray


if __name__ == '__main__':

    parser = argparse.ArgumentParser(
        description='Yield distances for next R farthest votings.')

    parser.add_argument('N', type=int, help='Number of voters')
    parser.add_argument('M', type=int, help='Number of candidates')
    parser.add_argument('R', type=int, help='Number of new votings to yield')
    parser.add_argument('--verbose', type=int,
                        help='Verbose level', default=1)
    parser.add_argument('--algorithm', type=str, help='Algorithm to use',
                        default='bh', choices=['bh', 'gurobi'])
    parser.add_argument('--json_output', type=bool,
                        help='Output as JSON', default=False)
    parser.add_argument('--load_hists', type=str,
                        help='Load votings from file')

    args = parser.parse_args()

    N = args.N
    M = args.M
    R = args.R
    votings_hists = [np.array([0] * M), np.array([N] * M)]
    R_Init = len(votings_hists)

    if args.verbose >= 1:
        print('r,dist,dist_prop,time')
    if args.json_output:
        print('{"hists":[')
        for voting in votings_hists:
            print(f'\t{list(voting)},')
    max_expected_dist = N * M // 2
    for i in range(R_Init + 1, R + R_Init + 1):
        start = time.time()
        if args.algorithm == 'bh':
            x, dist = bh.basin_hopping(
                votings_hists, N, niter=N*M*2, seed=2137)
        elif args.algorithm == 'gurobi':
            x, dist = gurobi.gurobi_ilp(votings_hists, N)
        dt = time.time() - start

        if args.verbose >= 1:
            print(f'{i},{dist},{dist/(N*M):.4f},{dt:.4f}')
        if args.json_output:
            print(f'\t{x}', end=',\n' if i < R + R_Init else '\n')

        max_expected_dist = dist
        votings_hists.append(x)
    if args.json_output:
        print(']}')
